### DISCUSSION TITLE
Question: When updating signal in parent component, child components don't update.

### DISCUSSION THREAD

**Discussion by TheFallenSpirit** [solidjs/solid#1402] [Q&A]
Here is the code I am using in my App file:
```tsx
const [loaded, setLoaded] = createSignal(false);
const RR: Component<ParentProps> = ({ children }) => (<RestrictedRoute loaded={loaded()}>{children}</RestrictedRoute>);
```

Here is the code from the RestrictedRoute component:
```tsx
const RestrictedRoute: Component<ParentProps<{ loaded: boolean }>> = ({ loaded, children }) => {
	if (!cookie.get('token')) loaded = true;
	if (!loaded) return <Loader/>;
	if (loaded && !store.user) return <Navigate href={'/'}/>;
	return children;
};
```

When the component RestrictedRoute is first mounted, the loaded signal is false in the parent component. But once the parent component updates the signal 'loaded' to be true, RestrictedRoute does not update and render the children. It still shows the loader as if the signal is still set to false.

I have spent most of my time coding in React so this is a little confusing. Please let me know if there is a solution to my issue. Thanks!

**Comment by ryansolid** [comment]
Yes, don't destructure props. It causes them to be accessed outside of a tracking context. 

```js
const [loaded, setLoaded] = createSignal(false);
const RR: Component<ParentProps> = (props) => (<RestrictedRoute loaded={loaded()}>{props.children}</RestrictedRoute>);
```

We don't recommend early returns because the component only runs once. So a decision could never change.
```js
const RestrictedRoute: Component<ParentProps<{ loaded: boolean }>> = (props) => {
    const loaded = () => !cookie.get('token') || props.loaded;

    return <Switch fallback={props.children}>
      <Match when={!loaded()}><Loader /></Match>
      <Match when={loaded() && !store.user}><Navigate href='/' /></Match>
    </Switch>
};
```

Solid works different enough from React I recommend looking at the tutorials.

**Reply by skuridin** [reply]
@ryansolid do you see yourself getting tired of questions about destructuring one day and agreeing to compromise performance for developer experience?

You did an amazing job with Solid but I think stuff like this slows down the community growth.

It doesn't have to be similar to React. What matters is an experience where it is extremely difficult to make a mistake. Right now developers, TypeScript, IDE have no clue that destructuring breaks reactivity.

**Reply by atk** [reply]
That's why the documentation is being rewritten to get that point across. Also, there's an eslint plugin, but it is not yet part of the official templates.

**Reply by strokirk** [reply]
Documentation will not be enough, I think. I’ve built a number of SolidJS apps but always forget that components have these footguns, leading to a lot of unnecessary debugging.

**Reply by ryansolid** [reply]
> @ryansolid do you see yourself getting tired of questions about destructuring one day and agreeing to compromise performance for developer experience?

What is the point then? The destructuring thing is because of the model. Short of using a compiler to rewrite your code that's what will happen. There is no performance to give up etc,.. because it is how it works. It works that way in Vue, Qwik, and basically any system that uses reactivity. It's a limitation of JavaScript language. So the solution other than getting back into VDOM or dirty checking territory is to give up JavaScript more or less.

**Reply by skuridin** [reply]
> So the solution other than getting back into VDOM

I was thinking in a different direction. If props destructuring breaks stuff, then make it impossible to destructure.  Maybe you can replace `props` object with `getProp` function (or Map?), like in the example below.
```jsx
const App = () => {
  const [count, setCount] = createSignal(0);
  return (
    <>
      <Counter count={count()} />
      <button onClick={() => setCount(count() + 1)}>Increment</button>
    </>
  )
}

const Counter = (getProp) => {
  const count = getProp('count')
  return <h1>{count}</h1>
}
```
Of course, no one wants breaking changes, so maybe you can do something similar to what Qwik does with their `component$` helper?

```jsx
const Counter = reactiveComponent$((getProp) => {
  const count = getProp('count')
  return <h1>{count}</h1>
})
```

I'm not telling that this is THE solution. My point is in preventing devs from using those parts of JS that conflict with Solid.

**Reply by kakserpom** [reply]
IMHO, the approach of SolidJS is easier to comprehend and more straightforward than that of React. The only thing to keep in mind is that a component function is executed only once. onMount() is only necessary when you need to run a piece of code after the initial rendering, otherwise you can just write your code inside of your component function.

**Reply by strokirk** [reply]
@ryansolid But we’re already using a compiler, are we not? The JSX is getting transformed into a fairly complex bit of code, could we not do that for the component arguments as well?

**Reply by skuridin** [reply]
@strokirk you can use Solid right in the browser without any pre-compilation.  
https://www.solidjs.com/guides/getting-started#buildless-options

**Reply by ryansolid** [reply]
@strokirk yeah. You essentially are suggesting a proxy that returns accessors on property access. A runtime wrapper written in user space could achieve this today.

In general it feels like extra but it definitely is an option. I think this sort of consideration might need to consider stuff like store design and affordances around proxied reactivity in general.
